### ISSUE TITLE
torch.nn.AdaptiveAvgPool2d  results in reproduction

### ISSUE THREAD

**Issue by konioyxgq** [pytorch/pytorch#49314]
Using **torch.nn.AdaptiveAvgPool2d** in the network results in no reproduction, and my version is '1.7.0a0+8deb4fe'
Why is that?

cc @albanD @mruberry @jbschlosser @heitorschueroff

**Comment by heitorschueroff** [comment]
Could you elaborate more on what you mean by no reproduction?

**Comment by konioyxgq** [comment]
Well, in order to be able to make recurrence of training process, I used **torch.backends.cudnn.benchmark=false**
    **torch.backends.cudnn.deterministic = True** and set seed. But when I used **torch.nn.AdaptiveAvgPool2d** in the network, I found that the training process could not be reproduction, and I replaced the **torch.nn.AdaptiveAvgPool2d** with **AvgPool2d**, which could be reproduction, so I suspected it was torch.nn.AdaptiveAvgPool2d's problem. The network I'm using RESnet_se. You can determine whether you can reproduction by comparing the weights in the model files saved by the two trainings. 
my version is '1.7.0a0+8deb4fe' and use A100 GPU

**Comment by mruberry** [comment]
Thank you for elaborating, @konioyxgq. I think I understand your issue. We added a new feature in PyTorch 1.7 that can help you understand this better, see here:

https://pytorch.org/docs/master/generated/torch.set_deterministic.html#torch.set_deterministic

This experimental feature lets you set a flag that will throw a runtime error when a nondeterministic operation is called. `torch.nn.AdaptiveAvgPool2d` is a nondeterministic operation when its input requires grad, as the documentation shows.

So this nondeterminism is expected and documented. If you like you could file an issue requesting a deterministic implementation of `torch.nn.AdaptiveAvgPool2d`. A deterministic version of the operation might be much slower, however.

**Comment by konioyxgq** [comment]
Thank you very much for your reply! I looked at the connection you gave me, my cuda version is 11.2, I used torch.mm() in the network, but I didn't set CUBLAS_WORKSPACE_CONFIG:4096:8 or CUBLAS_WORKSPACE_CONFIG=16:8. In this case, I compared the model saved by the two trainings at epoch 1, and the model parameters are the same. Is it going to be different in later training because I didn't set CUBLAS_WORKSPACE_CONFIG?

**Comment by mruberry** [comment]
> Thank you very much for your reply! I looked at the connection you gave me, my cuda version is 11.2, I used torch.mm() in the network, but I didn't set CUBLAS_WORKSPACE_CONFIG:4096:8 or CUBLAS_WORKSPACE_CONFIG=16:8. In this case, I compared the model saved by the two trainings at epoch 1, and the model parameters are the same. Is it going to be different in later training because I didn't set CUBLAS_WORKSPACE_CONFIG?

That's a good question, and I believe the answer is "maybe." cc @kurtamohler who can likely give you a better answer.

**Comment by konioyxgq** [comment]
Thank you very much @mruberry , excuse me @kurtamohler

**Comment by kurtamohler** [comment]
I'm not sure if there is a better answer than "maybe". Probably the only way to find out if your model would have been different with `CUBLAS_WORKSPACE_CONFIG` is to actually retrain it.

**Comment by mruberry** [comment]
Thanks @kurtamohler. @konioyxgq, I believe this issue is closed. If that's not correct please reopen it. As mentioned above, if you'd like you may file a separate issue requesting a deterministic torch.nn.AdaptiveAvgPool2d.
